r1cs/gadgets/std/src/groups/curves/short_weierstrass/short_weierstrass_jacobian.rs

```diff
@@ -114,14 +114,145 @@ impl<P, ConstraintF, F> AffineGadget<P, ConstraintF, F>
 
     #[inline]
     /// Incomplete, unsafe, addition: neither `self` nor `other` can be the neutral
-    /// element.
+    /// element, and other != ±self.
     pub fn add_unsafe<CS: ConstraintSystem<ConstraintF>>(
         &self,
         cs: CS,
         other: &Self,
     ) -> Result<Self, SynthesisError> {
         self.add_internal(cs, other, false)
     }
+
+    #[inline]
+    /// Compute 2 * self + other as (self + other) + self: this requires less constraints
+    /// than computing self.double().add(other).
+    /// Neither `self` nor `other` can be the neutral element, and other != ±self;
```

Comment:
And `(self + other) =! +/- self`, which under our presumption that `other != 0` is equivalent to `2*self + other != 0`.